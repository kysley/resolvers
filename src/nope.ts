import { appendErrors, transformToNestObject, Resolver } from 'react-hook-form';
import NopeObject from 'nope-validator/lib/umd/NopeObject';

const parseNopeSchema = (error: any, validateAllFieldCriteria: boolean) => {
  const errors = Object.keys(error);

  return Array.isArray(errors)
    ? errors.reduce((acc: Record<string, any>, fieldName) => {
        return {
          ...acc,
          ...(fieldName
            ? acc[fieldName] && validateAllFieldCriteria
              ? {
                  [fieldName]: appendErrors(
                    fieldName,
                    validateAllFieldCriteria,
                    acc,
                    '',
                    error[fieldName],
                  ),
                }
              : {
                  [fieldName]: acc[fieldName] || {
                    message: error[fieldName],
                    ...(validateAllFieldCriteria
                      ? {
                          types: {
                            '': error[fieldName] || true,
                          },
                        }
                      : {}),
                  },
                }
            : {}),
        };
      }, {})
    : [];
};

export const nopeResolver = <TFieldValues extends Record<string, any>>(
  schema: NopeObject,
): Resolver<TFieldValues> => async (
  values,
  _,
  validateAllFieldCriteria = false,
) => {
  const error = schema.validate(values);
  if (error) {
    return {
      values: {},
      errors: transformToNestObject(
        parseNopeSchema(error, validateAllFieldCriteria),
      ),
    };
  } else {
    return {
      values: values as any,
      errors: {},
    };
  }
};
