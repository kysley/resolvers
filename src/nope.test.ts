import Nope from 'nope-validator';
import { nopeResolver } from './nope';

const Article = Nope.object().shape({
  id: Nope.number().required(),
  title: Nope.string().required(),
  isPublished: Nope.string(),
  tags: Nope.array().of(Nope.string()).required(),
  author: Nope.object().shape({
    id: Nope.number().required(),
  }),
});

describe('nopeResolver', () => {
  it('should return correct value', async () => {
    const data = {
      id: 2,
      title: 'test',
      tags: ['news', 'features'],
      author: {
        id: 1,
      },
    };
    expect(await nopeResolver(Article)(data)).toEqual({
      values: data,
      errors: {},
    });
  });

  it('should return errors', async () => {
    const data = {
      id: '2',
      title: 2,
      tags: [2, 3],
      author: {
        id: 'test',
      },
    };
    expect(await nopeResolver(Article)(data)).toMatchSnapshot();
  });
});
