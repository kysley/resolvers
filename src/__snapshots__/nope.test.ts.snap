// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nopeResolver should return errors 1`] = `
Object {
  "errors": Object {
    "tags": Object {
      "message": "One or more elements are of invalid type",
    },
  },
  "values": Object {},
}
`;
